File: python/linked_list.py
from node import Node


class LinkedList:
    """Linked List implementation with new node appended to end"""

    def __init__(self):
        self.head = None

    def insert(self, value):
        """Insert node with value at end of linked list"""
        node = Node(value)
        if self.head is None:
            self.head = node
            return
        current_node = self.head
        while current_node.next != None:
            current_node = current_node.next
        current_node.next = node

    def is_empty(self):
        """ "Returns True if the linked list is empty, else False"""
        if self.head is None:
            return True
        return False

    def search(self, value):
        """Returns the index of a value in the linked list"""
        index = 0
        current = self.head
        while current != None:
            if current.value == value:
                return index
            current = current.next
            index += 1
        return -1

    def view_list(self):
        """prints each value in the linked list"""
        current = self.head
        while current is not None:
            print(current.value)
            current = current.next

    def remove_by_value(self, value, all=False):
        """Removes a node by its value, will remove all nodes with value if second parameter is set to True"""
        if self.head is not None and self.head.value == value:
            self.head = self.head.next
            if not all:
                return
        current = self.head
        while current is not None and current.next is not None:
            if current.next.value == value:
                current.next = current.next.next
                if not all:
                    return
            current = current.next
        return -1

    def remove_by_index(self, index):
        """Removes a node by its index"""
        if self.head is not None and index == 0:
            node_value = self.head.value
            self.head = self.head.next
            return node_value
        current = self.head
        idx = 1
        while current is not None:
            if idx == index:
                node_value = current.next.value
                current.next = current.next.next
                return node_value
            current = current.next
            idx += 1
        return -1

    def size(self):
        """Returns the size of a list"""
        index = 0
        current = self.head
        while current != None:
            index += 1
            current = current.next
        return index


if __name__ == "__main__":
    print("Add new node to back")
    new_list = LinkedList()
    print(new_list.is_empty())  # True
    new_list.insert(5)
    print(new_list.is_empty())  # False
    new_list.insert(10)
    new_list.insert(15)
    new_list.insert(25)
    new_list.insert(35)
    new_list.insert(25)
    new_list.view_list()  # print numbers 5, 10, 15, 25, 35, 25
    print()
    print(new_list.remove_by_index(2))  # 15
    print()
    new_list.view_list()  # print numbers 5, 10, 25, 35, 25
    print()
    print(new_list.search(25))  # 2
    print(new_list.search(40))  # -1
    print()
    new_list.remove_by_value(25, all=True)
    new_list.view_list()  # print numbers 5, 10, 35
    print(new_list.size())  # 3
    print(new_list.remove_by_value(50))  # -1
    print(new_list.remove_by_index(10))  # -1

File: js/linkedList.js
const {Node} = require("./node.js")

class LinkedList {
    constructor(){
        this.head  = null
        this.tail = null
        this.length = 0
    }

    push(value){
        let node = new Node(value)
        if (this.head == null) {
           this.head = node
           this.tail = node
        }
        else {
            this.tail.next =  node;
            this.tail = node
        }
        this.length++;
        return this
    }
    pop(){
        if (this.head == null) return undefined;
        let current = this.head;
        let newTail = current
        while (current.next){
            newTail = current
            current = current.next
        }

        this.tail = newTail
        this.tail.next = null
        this.length--;
        if(this.length === 0){
            this.head = null;
            this.tail = null;
        }
        return current

    }
    shift(){
        if (!this.head) return null;
        let currentHead = this.head;
        this.head = currentHead.next
        currentHead.next = null
        this.length--;
        if (this.length === 0){
            this.tail = null;
        }
        return currentHead

    }
    unshift(value){
        let node = new Node(value);
        if (!this.head){
            this.head = node;
            this.tail = this.head;
        } else {
        node.next = this.head;
        this.head = node
        }
        this.length++;
        
    }
    get(i){
        if (i >= this.length || i < 0) return null;
        let current = this.head;
        for (let j = 0; j < i; j++){
            current = current.next
        }
        return current
    }
    set(i, value){
        let foundNode = this.get(i);
        if (foundNode){
            foundNode.value = value;
            return true;
        }
        return false;
    }
    insert(i, value){
        if (i < 0 || i > this.length) return undefined;
        else if (i === 0) this.unshift(value);
        else if (i === this.length) this.push(value);
        else {
            let node = new Node(value);
            let prevNode = this.get(i-1);
            let temp = prevNode.next
            prevNode.next = node
            node.next = temp
            this.length++;
        }
    }
    remove(i){
        if (i < 0 || i > this.length) return undefined;
        if (i === 0) return this.shift();
        if (i === this.length) return this.pop();
        let prevNode = this.get(i-1)
        let removed = prevNode.next
        prevNode.next = prevNode.next.next
        this.length--;
        removed.next = null
        return removed;
    }
    reverse(){
        if (this.head == null) return null;
        let node = this.head
        this.head = this.tail
        this.tail = node
        let prev = null;
        let next;
        for (let i = 0; i < this.length; i++){
            next = node.next;
            node.next = prev;
            prev = node;
            node = next;
        }
    }
    static insertNodes(values, linkedList){
        for (let i = 0; i < values.length; i++){
            linkedList.push(values[i])
        }
    }
}

new_list = new LinkedList();
LinkedList.insertNodes([1,2,3,4,5,6,7,8,9,10], new_list)
new_list.unshift(100)
new_list.unshift(200)
new_list.reverse();
console.log(new_list)

